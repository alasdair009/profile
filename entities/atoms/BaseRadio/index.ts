export { BaseRadio } from "./BaseRadio";
export type { BaseRadioProps } from "./BaseRadio";
