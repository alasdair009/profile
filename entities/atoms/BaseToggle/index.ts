export { BaseToggle } from "./BaseToggle";
export type { BaseToggleProps } from "./BaseToggle";
