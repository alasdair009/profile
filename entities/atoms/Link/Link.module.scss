@use "../../../app/styles/colors";
@use "../../../app/styles/animation";
@use "../../../app/styles/dimensions";
@use "../../../app/styles/typography";
@use "../../../app/styles/lineClamp";
@use "../../../app/styles/helpers/rem";

@function generatePolygonLine($yOffset: 0) {
  @return polygon(
    0 0,
    30% #{$yOffset * 0.9%},
    50% #{$yOffset * 1%},
    70% #{$yOffset * 0.9%},
    100% 0,
    100% 25%,
    70% #{25 + $yOffset * 0.9%},
    50% #{25 + $yOffset * 1%},
    30% #{25 + $yOffset * 0.9%},
    0 25%
  );
}

@keyframes pressDown {
  from {
    clip-path: generatePolygonLine(0);
  }
  to {
    clip-path: generatePolygonLine(50);
  }
}

@keyframes release {
  0% {
    clip-path: generatePolygonLine(50);
  }
  60% {
    clip-path: generatePolygonLine(-30);
  }
  80% {
    clip-path: generatePolygonLine(20);
  }
  100% {
    clip-path: generatePolygonLine(0);
  }
}

.root {
  align-items: center;
  color: colors.$colorWhiteGhost;
  display: inline-flex;
  font-size: inherit;
  font-weight: map-get(typography.$fontWeights, bold);
  max-width: rem.toRem(map-get(dimensions.$sizes, s512));
  position: relative;
  text-decoration: none;
  word-break: break-word;

  &:hover {
    filter: brightness(1.3);

    &::after {
      animation: pressDown animation.animationDuration(0.1);
      animation-fill-mode: forwards;
    }

    .hoverIFrameWrapper {
      display: block;
    }
  }

  &:visited,
  &:active {
    color: colors.$colorWhiteGhost;
  }

  &::after {
    animation: release animation.animationDuration(0.2);
    animation-fill-mode: forwards;
    background: colors.$colorWhiteGhost;
    bottom: calc(#{rem.toRem(7)} * -1);
    clip-path: generatePolygonLine();
    content: "";
    height: rem.toRem(map-get(dimensions.$sizes, s8));
    left: 0;
    position: absolute;
    width: 100%;
  }

  &Clamp {
    @include lineClamp.lineClamp;
  }

  &.Large {
    color: colors.$colorGreenGrass;
    font-size: rem.toRem(map-get(typography.$fontSizes, large));
    padding-bottom: calc(rem.toRem(map-get(dimensions.$sizes, s16)) + 2px);
    text-transform: uppercase;

    &:visited,
    &:active {
      color: colors.$colorGreenGrass;
    }

    &::after {
      bottom: 0;
      height: rem.toRem(map-get(dimensions.$sizes, s16));
    }
  }
}

.newTabIcon {
  display: inline-block;
  height: rem.toRem(map-get(dimensions.$sizes, s16));
  margin-left: rem.toRem(map-get(dimensions.$sizes, s4));
  width: rem.toRem(map-get(dimensions.$sizes, s16));
}

.hoverIFrameWrapper {
  aspect-ratio: 16 / 9;
  background: colors.$colorGreyDark;
  border: rem.toRem(map-get(dimensions.$sizes, s8)) solid colors.$colorGreyDark;
  box-shadow: 0 0 rem.toRem(2) rem.toRem(2) colors.$colorblackEvil;
  display: none;
  height: calc(
    rem.toRem(map-get(dimensions.$sizes, s128)) +
      rem.toRem(map-get(dimensions.$sizes, s16))
  );
  overflow: hidden;
  position: absolute;
  left: 50%;
  top: 100%;
  transform: translate(-50%, 0);
  transition: all animation.animationDuration(0.2);
  user-select: none;
}

.hoverIFrame {
  aspect-ratio: unset;
  height: calc(1000% - rem.toRem(map-get(dimensions.$sizes, s16)));
  left: 50%;
  position: absolute;
  top: 50%;
  transform: translate(-50%, -50%) scale(0.5);
  width: calc(1000% - rem.toRem(map-get(dimensions.$sizes, s16)));
}
