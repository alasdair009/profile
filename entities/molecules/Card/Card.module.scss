@use "../../../app/styles/colors";
@use "../../../app/styles/dimensions";
@use "../../../app/styles/helpers/rem";
@use "sass:map";

.root {
  background: colors.$colorGreyDark;
  box-shadow: 0 0 rem.toRem(10) rem.toRem(6) rgba(colors.$colorWhiteGhost, 0.35);
  display: flex;
  height: rem.toRem(300);
  margin: 0 auto;
  padding: dimensions.getSize(s8);
  position: relative;
  width: 100%;

  @media (map.get(dimensions.$device, xsmall)) {
    max-width: dimensions.getSize(s256);
  }
}

.cardLink {
  align-items: center;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  overflow: hidden;
  width: 100%;

  &::after {
    display: none;
  }

  &:hover {
    filter: brightness(1.3);
  }
}

.figure {
  display: flex;
  margin: 0;
  position: relative;
  width: 100%;
}

.cardImage {
  align-content: center;
  background: colors.$colorblackEvil;
  display: flex;
  min-height: dimensions.getSize(s128);
  justify-content: center;
  object-fit: cover;
  text-align: center;
  width: 100%;

  @media (map.get(dimensions.$device, xsmall)) {
    min-height: dimensions.getSize(s96);
  }
}

.date {
  color: colors.$colorWhiteGhost;
  display: block;
  font-style: italic;
  text-align: center;
}
