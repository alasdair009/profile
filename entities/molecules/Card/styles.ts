"use client";
import styled from "styled-components";
import { colors, Link, sizes } from "@/entities";
import { rem } from "polished";
import Image from "next/image";

export const Root = styled.article`
  background: ${colors.greyDark};
  box-shadow: 0 0 ${rem(3)} ${rem(3)} ${colors.blackEvil};
  display: flex;
  height: ${rem(300)};
  margin: 0 auto;
  max-width: ${sizes.s256.rem};
  padding: ${sizes.s8.rem};
  position: relative;
  width: 100%;
`;

export const CardLink = styled(Link)`
  align-items: center;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  overflow: hidden;
  width: 100%;

  &::after {
    display: none;
  }

  &:hover {
    filter: brightness(1.3);
  }
`;
export const Figure = styled.figure`
  display: flex;
  margin: 0;
  width: 100%;
`;

export const CardImage = styled(Image)`
  align-content: center;
  background: ${colors.blackEvil};
  display: flex;
  min-height: ${sizes.s96.rem};
  justify-content: center;
  object-fit: contain;
  text-align: center;
  width: 100%;
`;

export const Content = styled.div``;

export const Date = styled.time`
    color: ${colors.whiteGhost};
    display: block;
    font-style: italic;
    text-align: center;
`;
