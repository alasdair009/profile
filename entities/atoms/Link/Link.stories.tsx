import { Link } from "./Link";
import { Meta, StoryObj } from "@storybook/react";

const meta: Meta<typeof Link> = {
  component: Link,
  argTypes: {
    children: {
      control: {
        type: "text",
      },
    },
  },
  args: {
    href: "#",
    children: "Go to page",
  },
};
export default meta;

export const Default: StoryObj<typeof Link> = {};
