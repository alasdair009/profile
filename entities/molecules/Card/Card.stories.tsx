import { Meta, StoryObj } from "@storybook/nextjs";
import { Card } from "./Card";
import { expect, within } from "storybook/test";
import { Link } from "../../atoms/Link";

const meta: Meta<typeof Card> = {
  component: Card,
  args: {
    title: "Article title",
    date: new Date("1 January 1980"),
    href: "#",
    imageAlt: "Image alt text",
  },
};
export default meta;

export const Default: StoryObj<typeof Card> = {
  play: async ({ args, canvasElement }) => {
    const canvas = within(canvasElement);
    const cardElement = canvas.getByTestId(Card.name);

    await expect(cardElement).toBeInTheDocument();
    await expect(canvas.getByText(args.title)).toBeInTheDocument();
    await expect(canvas.getByText("1 Jan 1980")).toBeInTheDocument();
    await expect(canvas.getByTestId(`${Card.name}Link`)).toHaveAttribute(
      "href",
      "#"
    );
    await expect(canvas.getByAltText(`${args.imageAlt}`)).toBeInTheDocument();
  },
};
