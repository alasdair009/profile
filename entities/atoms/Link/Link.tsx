import { NewTabIcon, StaticLink, TransitionLink } from "./styles";
import type { LinkProps } from "./types";
import { isExternalDomain } from "@/lib/domains";
import newTabIcon from "../../assets/new-tab-icon.svg";

/**
 * Link to a new path or url.
 */
export function Link({
  variant = "regular",
  children,
  href,
  ...rest
}: LinkProps) {
  const isExternalURL = isExternalDomain(href);

  if (process.env.ALLOW_TRANSITIONS) {
    return (
      <TransitionLink
        $variant={variant}
        href={href}
        target={isExternalURL ? "_blank" : "_self"}
        rel={isExternalURL ? "noopener" : undefined}
        data-testid={Link.name}
        {...rest}
      >
        {children ? children : href}
        {isExternalURL && (
          <NewTabIcon
            src={newTabIcon}
            title="Third party site, opens in a new tab."
            alt="Icon to represent this opens in a new tab"
          />
        )}
      </TransitionLink>
    );
  }

  return (
    <StaticLink
      $variant={variant}
      href={href}
      target={isExternalURL ? "_blank" : "_self"}
      rel={isExternalURL ? "noopener" : undefined}
      data-testid={Link.name}
      {...rest}
    >
      {children ? children : href}
      {isExternalURL && (
        <NewTabIcon
          src={newTabIcon}
          title="Third party site, opens in a new tab."
          alt="Icon to represent this opens in a new tab"
        />
      )}
    </StaticLink>
  );
}
