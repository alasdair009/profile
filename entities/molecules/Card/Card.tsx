import { CardImage, CardLink, Content, Date, Figure, Root } from "./styles";
import { HTMLAttributes } from "react";
import { Heading } from "@/entities";
import { StaticImageData } from "next/image";
import amLogo from "../../assets/am.svg";

type CardProps = {
  href: string;
  title: string;
  image?: StaticImageData;
  imageAlt?: string;
  date: Date;
} & HTMLAttributes<HTMLDivElement>;

/**
 * Box with image and copy that links to another page
 */
export function Card({
  href,
  title,
  image = amLogo,
  imageAlt = "AM Logo",
  date,
  ...rest
}: CardProps) {
  return (
    <Root {...rest}>
      <CardLink href={href}>
        <Figure>
          <CardImage src={image} alt={imageAlt} fill={true} />
        </Figure>
        <Content>
          <Heading level="h3" as="h2" align="center">
            {title}
          </Heading>
          <Date dateTime={date.toISOString()}>
            {date.toLocaleDateString("en-GB", {
              year: "numeric",
              month: "short",
              day: "numeric",
            })}
          </Date>
        </Content>
      </CardLink>
    </Root>
  );
}
