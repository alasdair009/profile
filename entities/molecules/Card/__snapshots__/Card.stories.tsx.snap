// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`molecules/Card Default play-test 1`] = `
<article data-testid="Card"
         class="sc-eowDPD fZITNq"
>
  <a target="_self"
     data-testid="Link"
     class="sc-eEPDDI gughtU sc-iAlELC bmQRoO"
     href="#"
  >
    <figure class="sc-oeqTF coCVHF">
      <img alt="Image alt text"
           loading="lazy"
           decoding="async"
           data-nimg="fill"
           class="sc-ezTrPE jyxelm"
           sizes="100vw"
           srcset="/static/media/entities/assets/am.svg?w=640&amp;q=75 640w, /static/media/entities/assets/am.svg?w=750&amp;q=75 750w, /static/media/entities/assets/am.svg?w=828&amp;q=75 828w, /static/media/entities/assets/am.svg?w=1080&amp;q=75 1080w, /static/media/entities/assets/am.svg?w=1200&amp;q=75 1200w, /static/media/entities/assets/am.svg?w=1920&amp;q=75 1920w, /static/media/entities/assets/am.svg?w=2048&amp;q=75 2048w, /static/media/entities/assets/am.svg?w=3840&amp;q=75 3840w"
           src="/static/media/entities/assets/am.svg?w=3840&amp;q=75"
           style="position: absolute; height: 100%; width: 100%; inset: 0px; color: transparent;"
      >
    </figure>
    <div class="sc-drnuxz">
      <h2 data-testid="Heading"
          class="sc-hfvVTD uqbXt"
      >
        Article title
      </h2>
      <time datetime="1980-01-01T00:00:00.000Z"
            class="sc-hDcvty WhVSS"
      >
        1 Jan 1980
      </time>
    </div>
  </a>
</article>
`;
